### PR TITLE
feat: implement bootstrap in meroctl

### PR DESCRIPTION
**Pull request by MatejVukosav**
TODO v2
- remove manual sleep and wait for actual action to finish

Test
`cargo run -p meroctl -- --node-name node1 --home data bootstrap start --merod-path {path} --app-path {path}`
or with app download
`cargo run -p meroctl -- --node-name node1 --home data bootstrap start --merod-path {path}`
If user didn't provide app-path, instructions at the bottom of log are provided. Otherwise not.

Log
```
Starting bootstrap process
Downloading demo app...
Demo app downloaded successfully.
Initializing node "node1"
Starting node "node1".
Node "node1" started successfully.
Creating context in "node1"
{"data":{"applicationId":"QhbtK4fe25jCbFLFDxBMEe5YrHBeHYd3ynQxjzZ8yzf"}}
{"data":{"contextId":"AdZmi9s5bZhW5be8TndLV51ZM5dxBWrjUKxSRM4Z3GJr","memberPublicKey":"3iWqdX7aSfkJumFepbuGxAL7hZFYiFKrwsLKZnX9HV1j"}}
Context created: "AdZmi9s5bZhW5be8TndLV51ZM5dxBWrjUKxSRM4Z3GJr"
Initializing node "node2"
Starting node "node2".
Node "node2" started successfully.
Inviting node "node2" to context "AdZmi9s5bZhW5be8TndLV51ZM5dxBWrjUKxSRM4Z3GJr"
{"data":"223oXn7wzDFF7LS76Z5vbqgGZn4bea8hVGXUaBPkwc8ZHxs5F26zFBeqUpTisCEaGshfSL37M9QqThw23fpox8Xpy1sgbHvoB6tQoX547BSEHBYDgUhBayvfVp6uqLx1qrRksxXJQa7prbV9fSH3Y3wCYVtSNKzBvb"}
Node "node2" successfully invited.
Joining node "node2" to context.
{"data":{"contextId":"AdZmi9s5bZhW5be8TndLV51ZM5dxBWrjUKxSRM4Z3GJr","memberPublicKey":"DREcrMahB4EA5oytkdsntejuXP5YMnNDPE3NTPFFYzNs"}}
Node "node2" joined successfully.
************************************************
🚀 Bootstrap finished. Nodes are ready to use! 🚀
Context id is "AdZmi9s5bZhW5be8TndLV51ZM5dxBWrjUKxSRM4Z3GJr" 
Connect to the node from https://calimero-network.github.io/core-app-template/
Open application in two separate windows to use it with two different nodes.
Application setup screen requires application id and node url.
Application id is "QhbtK4fe25jCbFLFDxBMEe5YrHBeHYd3ynQxjzZ8yzf" 
Node "node1" url is http://localhost:2428
Node "node1" url is http://localhost:2429
************************************************
```

### CODE REVIEW

File: crates/merod/src/cli/init.rs [hunk 1]
```diff
@@ -157,7 +157,7 @@ impl InitCommand {
                 }
             }
             if !self.force {
-                bail!("Node is already initialized in {:?}", path);
+                warn!("Node is already initialized in {:?}", path);
```
Comment (MatejVukosav) [inline crates/merod/src/cli/init.rs:160]: Not sure why was this error before and not just warning.

File: crates/meroctl/src/cli/context/join.rs [hunk 2]
```diff
@@ -38,22 +38,39 @@ impl Report for JoinContextResponse {
...
         .await?;
 
         environment.output.write(&response);
 
+        let join_result = response.data;
+
+        let join_result = join_result
+            .ok_or_else(|| eyre::eyre!("No invitation payload found in the response"))?;
+
+        println!("Join result {:?}", join_result);
```
Comment (MatejVukosav) [inline crates/meroctl/src/cli/context/join.rs]:
```suggestion
```

File: crates/merod/src/cli/init.rs (hunk 1 above)
Comment (miraclx) [reply crates/merod/src/cli/init.rs:160]: good catch

File: crates/meroctl/src/cli/app/install.rs [hunk 3]
```diff
@@ -36,20 +37,35 @@ impl Report for InstallApplicationResponse {
...
+        Ok(())
+    }
+
+    pub async fn install_app(
+        path: Option<Utf8PathBuf>,
+        hash: Option<Hash>,
+        metadata: Option<String>,
+        url: Option<String>,
+        environment: &Environment,
+    ) -> Result<ApplicationId> {
```
Comment (miraclx) [inline crates/meroctl/src/cli/app/install.rs]:
```suggestion
        let _ignored = self.install_app(environment).await;

        Ok(())
    }

    pub async fn install_app(&self, environment: &Environment) -> Result<ApplicationId> {
```

File: crates/meroctl/src/cli/bootstrap/start.rs [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+        let config = load_config(&environment.args.home, &environment.args.node_name)?;
+        let multiaddr = fetch_multiaddr(&config)?;
+        let client = Client::new();
+
+        let app_hash = Some(Hash::new("hash".as_bytes()));
+        let app_metadata = Some("".to_owned());
+        let url = Some("".to_owned());
+
+        let application_id =
+            InstallCommand::install_app(app_path, app_hash, app_metadata, url, environment).await?;
```
Comment (miraclx) [inline crates/meroctl/src/cli/bootstrap/start.rs]:
```suggestion
        let command = InstallCommand {
            path: app_path,
            url: Some("".to_owned()),
            metadata: Some("".to_owned()),
            hash: Some(Hash::new("hash".as_bytes()))
        };

        let application_id = command.install_app(environment).await?;
```

**Comment by MatejVukosav** [comment]
Run tests locally and they are passing
`cargo build -p meroctl -p merod -p e2e-tests`

@fbozic did you saw this error before?

File: crates/meroctl/src/cli/bootstrap/start.rs [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+        )
+        .await?;
+
+        println!("************************************************");
+        println!("🚀 Bootstrap finished. Nodes are ready to use! 🚀");
+
+        if demo_app {
+            println!(
+                "Application is available at https://calimero-network.github.io/core-app-template/"
+            );
```
Comment (miraclx) [inline crates/meroctl/src/cli/bootstrap/start.rs]: "connect to this node from?" or something like that

File: crates/meroctl/src/cli/bootstrap/start.rs [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+                node1_environment.args.node_name,
+                application_id.to_string(),
+                node1_server_port
+            );
+            println!(
+                "Application id for node {:?} is {:?} and url is http://localhost:{}",
+                node1_environment.args.node_name,
+                application_id.to_string(),
+                node2_server_port
+            );
```
Comment (miraclx) [inline crates/meroctl/src/cli/bootstrap/start.rs]: why are we printing `application_id` for both nodes if it's the same?

the context id (and the node identity), which I'd think to be the more relevant bit, isn't also printed?

Comment (MatejVukosav) [reply crates/meroctl/src/cli/bootstrap/start.rs]: I followed fields in the UI. Thought it may be easier for them to copy those parts like that

Comment (MatejVukosav) [reply crates/meroctl/src/cli/bootstrap/start.rs]: They don’t need to write it anywhere and atm they will have only one context to select.

If they need it for some reason, it is printed during the bootstrap.

File: crates/meroctl/src/cli/bootstrap/start.rs (hunk 4 above)
Comment (MatejVukosav) [reply crates/meroctl/src/cli/bootstrap/start.rs:301]: no method named `install_app` found for struct `InstallCommand`

File: crates/merod/Cargo.toml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "merod"
-version = "0.2.1"
+version = "0.2.2"
```
Comment (xilosada) [inline crates/merod/Cargo.toml]: @MatejVukosav @miraclx Let's handle the release in a separate PR and include the release notes there. This way, we can incorporate any additional changes and avoid spamming releases unnecessarily.

File: crates/meroctl/src/cli/bootstrap.rs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+use eyre::Result as EyreResult;
+use start::StartBootstrapCommand;
+
+use super::Environment;
+
+mod start;
+
+pub const EXAMPLES: &str = r"
+  # Setup and run 2 nodes with demo app
+  $ meroctl -- --node-name node1 bootstrap start --merod-path /path/to/merod
```
Comment (frdomovic) [inline crates/meroctl/src/cli/bootstrap.rs:12]: Why do we only need to provide name for node1?

Comment (MatejVukosav) [reply crates/meroctl/src/cli/bootstrap.rs:12]: Only node_name does not have default value

File: crates/meroctl/src/cli/bootstrap/start.rs [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+
+        if demo_app {
+            println!(
+                "Connect to the node from https://calimero-network.github.io/core-app-template/"
+            );
+            println!(
+                "Open application in two separate windows to use it with two different nodes."
+            );
+            println!("Application setup screen requires application id and node url.");
+            println!("Application id is {:?} ", application_id.to_string(),);
```
Comment (miraclx) [inline crates/meroctl/src/cli/bootstrap/start.rs:129]: It also requires knowing the context ID, but that isn't printed here.

File: crates/meroctl/src/cli/context/join.rs [hunk 10]
```diff
@@ -38,6 +38,11 @@ impl Report for JoinContextResponse {
 
 impl JoinCommand {
     pub async fn run(self, environment: &Environment) -> EyreResult<()> {
+        let _ignored = self.join(environment).await?;
+        Ok(())
+    }
+
+    pub async fn join(self, environment: &Environment) -> EyreResult<()> {
```
Comment (miraclx) [inline crates/meroctl/src/cli/context/join.rs]: `join` isn't useful here.

**Review by miraclx — APPROVED** [review]

**Pull request merged**